lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2152,24 +2157,27 @@ listStakeKeys
     -> ctx
     -> ApiT WalletId
     -> Handler (ApiStakeKeys n)
-listStakeKeys lookupStakeRef ctx (ApiT wid) = do
+listStakeKeys lookupStakeRef ctx (ApiT wid) =
     withWorkerCtx ctx wid liftE liftE $ \wrk -> liftHandler $ do
-            (wal, meta, pending) <- W.readWallet @_ @s @k wrk wid
-            let utxo = availableUTxO @s pending wal
-
-            mourAccount <- fmap (fmap fst . eitherToMaybe)
-                <$> liftIO . runExceptT $ W.readRewardAccount @_ @s @k @n wrk wid
-            ourApiDelegation <- liftIO $ toApiWalletDelegation (meta ^. #delegation)
-                (unsafeExtendSafeZone (timeInterpreter $ ctx ^. networkLayer))
-            let ourKeys = case mourAccount of
-                    Just acc -> [(acc, 0, ourApiDelegation)]
-                    Nothing -> []
-
-            liftIO $ listStakeKeys' @n
-                utxo
-                lookupStakeRef
-                (fetchRewardAccountBalances nl)
-                ourKeys
+        (wal, meta, pending) <- W.readWallet @_ @s @k wrk wid
+        let utxo = availableUTxO @s pending wal
+
+        mourAccount <- fmap (fmap fst . eitherToMaybe)
+            <$> liftIO . runExceptT $ W.readRewardAccount @_ @s @k @n wrk wid
+        ourApiDelegation <- liftIO $ toApiWalletDelegation (meta ^. #delegation)
+            (unsafeExtendSafeZone (timeInterpreter $ ctx ^. networkLayer))
+
+        -- With ADP-808 we will be able to have many keys. Currently there's
+        -- only one.
+        let ourKeys = case mourAccount of
+                Just acc -> [(acc, 0, ourApiDelegation)]
+                Nothing -> []
+
+        liftIO $ listStakeKeys' @n
+            utxo
+            lookupStakeRef
+            (fetchRewardAccountBalances nl)
```

Comment:
I still don't like this way of splitting it because it means you still have a big function that mixes pure and monadic code.